std/internal/cstring.d

```diff
@@ -75,57 +82,99 @@ $(D strlen(str.tempCString()))). Incorrect usage of this function may
 lead to memory corruption.
 See $(RED WARNING) in $(B Examples) section.
 */
-auto tempCString(To = char, From)(in From[] str) nothrow @nogc
-if(isSomeChar!To && isSomeChar!From)
+
+auto tempCString(To = char, From)(From str)
+    if (isSomeChar!To && (isInputRange!From || isSomeString!From))
 {
-    import core.checkedint : addu;
     import core.exception : onOutOfMemoryError;
+    import core.stdc.stdlib;
+    import core.stdc.string : memcpy;
+
+    enum useStack = cast(To*) size_t.max;
 
-    enum useStack = cast(To*) -1;
+    alias CF = Unqual!(ElementEncodingType!From);
 
     static struct Res
     {
-    nothrow @nogc:
+      nothrow @nogc:
 
         @disable this();
         @disable this(this);
         alias ptr this;
 
         @property inout(To)* buffPtr() inout @safe pure
-        { return _ptr == useStack ? _buff.ptr : _ptr; }
+        {
+            return _ptr == useStack ? _buff.ptr : _ptr;
+        }
 
         @property const(To)* ptr() const @safe pure
-        { return buffPtr; }
+        {
+            return buffPtr;
+        }
 
         ~this()
-        { if(_ptr != useStack) rawFree(_ptr); }
+        {
+            if (_ptr != useStack)
+                core.stdc.stdlib.free(_ptr);
+        }
 
-    private:
+      private:
         To* _ptr;
-        To[256] _buff;
+        version (unittest)
+        {
+            To[16 / To.sizeof] _buff;   // smaller size to trigger reallocations
```

Comment:
This was removed (https://github.com/D-Programming-Language/phobos/pull/3410), superseded by https://github.com/D-Programming-Language/phobos/pull/3415, which doesn't have the same problem.

The issue you ran into is a codegen problem, as Kenji has identified.